pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -678,33 +704,102 @@ func (t *Test) runCommandFunc(nodes option.NodeListOption, cmd string) stepFunc
-	if err != nil {
-		return nil, err
+	skipVersions := t.prng.Float64() < t.options.skipVersionProbability
+	isAvailable := func(v *clusterupgrade.Version) bool {
+		if t.clusterArch() != vm.ArchARM64 {
+			return true
+		}
+
+		return v.AtLeast(minSupportedARM64Version)
 	}
 
-	if t.clusterArch() != vm.ArchARM64 {
-		return releases, nil
+	var numSkips int
+	// possiblePredecessorsFor returns a list of possible predecessors
+	// for the given release `v`. If skip-version is enabled and
+	// supported, this function will return both the immediate
+	// predecessor along with the predecessor's predecessor. This is the
+	// function to change in case the rules around what upgrades are
+	// possible in CRDB change.
+	possiblePredecessorsFor := func(v *clusterupgrade.Version) ([]*clusterupgrade.Version, error) {
+		pred, err := t.predecessorFunc(t.prng, v)
+		if err != nil {
+			return nil, err
+		}
+
+		if !isAvailable(pred) {
+			return nil, nil
+		}
+
+		// If skip-version upgrades are not enabled, the only possible
+		// predecessor is the immediate predecessor release.
+		if !skipVersions {
+			return []*clusterupgrade.Version{pred}, nil
+		}
+
+		predPred, err := t.predecessorFunc(t.prng, pred)
+		if err != nil {
+			return nil, err
+		}
+
+		if predPred.AtLeast(minSupportedSkipVersionUpgrade) {
+			// If the predecessor's predecessor supports skip-version
+			// upgrades and we haven't performed a skip-version upgrade yet,
+			// do it. This logic makes sure that, when skip-version upgrades
+			// are enabled, it happens when upgrading to the current
+			// release, which is the most important upgrade to be tested on
+			// any release branch.
+			if numSkips == 0 {
+				numSkips++
+				return []*clusterupgrade.Version{predPred}, nil
+			}
+
+			// If we already performed a skip-version upgrade on this test
+			// plan, we can choose to do another one or not.
+			return []*clusterupgrade.Version{pred, predPred}, nil
+		}
+
+		// If the predecessor is too old and does not support skip-version
+		// upgrades, it's the only possible predecessor.
+		return []*clusterupgrade.Version{pred}, nil
 	}
 
-	var armReleases []*clusterupgrade.Version
-	for _, r := range releases {
-		if r.AtLeast(minSupportedARM64Version) {
-			armReleases = append(armReleases, r)
+	currentVersion := clusterupgrade.CurrentVersion()
+	var upgradePath []*clusterupgrade.Version
+	numUpgrades := t.numUpgrades()
+
+	for j := 0; j < numUpgrades; j++ {
+		predecessors, err := possiblePredecessorsFor(currentVersion)
+		if err != nil {
+			return nil, err
 		}
+
+		// If there are no valid predecessors, it means some release is
+		// not available for the cluster architecture. We log a warning
+		// below in case we have a shorter upgrade path than requested
+		// because of this.
+		if len(predecessors) == 0 {
+			break
+		}
+
+		chosenPredecessor := predecessors[t.prng.Intn(len(predecessors))]
+		upgradePath = append(upgradePath, chosenPredecessor)
```

Comment:
Nit: might be worth asserting that it maintains a strict descending order; it's not immediately obvious,

```
assert(len(upgradePath) == 1 || ugradePath[len(upgradePath]-1] < chosenPredecessor`
```

Reply:
IMO, this invariant is part of the contract of the `releases` package (i.e., when you call `releases.LatestPredecessor`, you get an immediate predecessor). If we wanted to add an assertion, I'd rather assert ordering when loading the releases file (`parseReleases`).